### DISCUSSION TITLE
Ignoring stale result from the server

### DISCUSSION THREAD

**Discussion by alexkirsz** [urql-graphql/urql#2349] [Q&A]
Hey,

Is there a way to selectively ignore results from the server in graphcache? Our use case is that we sometimes have newer versions of a particular entity locally that are currently syncing to the server, but we receive an older version in the meantime. The newer version gets temporarily overwritten, and then updated back once we hear from the server again, which causes something of a rubberbanding effect.

In our case, all entities have an `updatedAt` field, so checking whether or not to update them is as simple as comparing `updatedAt`: if the local `updatedAt` is larger than the newly received `updatedAt`, we should ignore the result while we have yet to hear back from the result of the mutation that caused the optimistic update.

Cheers,
Alex

**Comment by saevarb** [comment]
I don't think you should be handling this with urql, but it's hard to tell from the information. 

Do you have a form which receives initial values, and then a query runs in the background and the form values are overwritten? In that case, that's 100% a form library thing, where you can usually specify that you don't want the form to reinitialize if the initial values change.

It could also sound like what is happening is the following:

1. You take entity and edit it in a form
2. You send updated entity to server with mutation
3. Leaving the form and, say, going to the page where it's displayed  displays stale entity because mutation isn't finished / the query isn't invalidated
4. The new entity arrives and and the display is updated

If that is the case, you probably want to look into optimistic updates. You can plug the updated entity straight into the cache immediately(so the new entity is displayed immediately) and it will then update transparently in the background, assuming your update logic is in order.

**Reply by alexkirsz** [reply]
Hey @saevarb, thanks for the answer. I am in fact using optimistic updates, I forgot to mention it in the original post.

What's happening is as follows:
1. We have an entity `E` with `updatedAt=0`, cached on the client and correctly synced with the server.
2. We have a query `Q` that fetches a list of entities and runs periodically. `Q` starts fetching.
3. While `Q` is fetching, we initiate mutation `M`, which optimistically mutates entity `E` in the cache with `updatedAt=1`.
4. Query `Q` finishes fetching and contains `E` with `updatedAt=0`. `E` is rolled back to its state at `updatedAt=0`.
5. Mutation `M` finishes fetching, and contains the new version of `E` with `updatedAt=1`. `E` is now `updatedAt=1` on both the client and the server.

In this flow, `E` sees `updatedAt=0`, then `updatedAt=1` (optimistic), then back to `updatedAt=0`, then back to `updatedAt=1`. I would like to be able to selectively ignore results about `E` from `Q` while I have an optimistic mutation running for `E`.

**Reply by saevarb** [reply]
I think the way to go here is to invalidate `Q` in the updater for `M`. If I understand the way `urql` works correctly, that will cancel the in-progress `Q` and (probably?) ensure it doesn't run until after the mutation is done. I think you can rely on the optimistic updater running synchronously and immediately.

**Reply by alexkirsz** [reply]
@saevarb Wouldn't that also always trigger `Q` to run after `M` finishes? In the example I gave, `Q` is running at the same time as `M`, but it is not always the case.
